### ISSUE TITLE
nccl 2.10.3 may have bug in multi data type in a group call

### ISSUE THREAD

**Issue by guo-ran** [NVIDIA/nccl#560]
my code run normally with nccl 2.9.8, but with nccl 2.10.3, get an error " an illegal memory access was encountered" after ncclAllReduce call
```
#include <cuda.h>
#include <cuda_fp16.h>
#include <nccl.h>
#include <iostream>
#include <vector>

void Check(cudaError_t err) {
  if (err != cudaSuccess) {
    std::cerr << cudaGetErrorString(err) << std::endl;
    exit(1);
  }
}

void Check(ncclResult_t err) {
  if (err != ncclSuccess) {
    std::cerr << ncclGetErrorString(err) << std::endl;
    exit(1);
  }
}

int main() {
  int elem_cnt = 4 * 1024 * 1024;
  int n_gpu = 2;
  std::vector<cudaStream_t> streams(n_gpu);
  std::vector<half *> half_buffers(n_gpu);
  std::vector<float *> float_buffers(n_gpu);
  std::vector<ncclComm_t> comms(n_gpu);
  for (int i = 0; i < n_gpu; ++i) {
    Check(cudaSetDevice(i));
    Check(cudaStreamCreate(&streams.at(i)));
    Check(cudaMalloc(&half_buffers.at(i), elem_cnt * sizeof(half)));
    Check(cudaMalloc(&float_buffers.at(i), elem_cnt * sizeof(float)));
    Check(cudaMemset(half_buffers.at(i), 0, elem_cnt * sizeof(half)));
    Check(cudaMemset(float_buffers.at(i), 0, elem_cnt * sizeof(float)));
  }

  Check(cudaDeviceSynchronize());
  ncclUniqueId unique_id;
  Check(ncclGetUniqueId(&unique_id));
  Check(ncclGroupStart());
  for (int i = 0; i < n_gpu; ++i) {
    Check(cudaSetDevice(i));
    Check(ncclCommInitRank(&comms.at(i), n_gpu, unique_id, i));
  }

  Check(ncclGroupEnd());
  Check(cudaDeviceSynchronize());

  Check(ncclGroupStart());
  for (int i = 0; i < n_gpu; i++) {
    Check(cudaSetDevice(i));
    Check(ncclAllReduce(float_buffers.at(i), float_buffers.at(i), elem_cnt,
                        ncclDataType_t::ncclFloat, ncclRedOp_t::ncclSum,
                        comms.at(i), streams.at(i)));
    Check(ncclAllReduce(half_buffers.at(i), half_buffers.at(i), elem_cnt,
                        ncclDataType_t::ncclFloat16, ncclRedOp_t::ncclSum,
                        comms.at(i), streams.at(i)));
  }
  Check(ncclGroupEnd());
  Check(cudaDeviceSynchronize());

  return 0;
}

```

**Comment by sjeaugey** [comment]
Thanks for the report. Can you give more info about the platform you are running on (server type, number of GPUs, type of GPUs, network used, ...)?

**Comment by jbachan** [comment]
Verified `ncclGroup()` with mixed datatype/element/collective could induce crash or corruption.

Fixed in 5f2f2f670f2604ff44d1f996733e218c0743af73
